linalg/operator.hpp

```diff
@@ -309,30 +309,37 @@ class Operator
 
 
 /// Base abstract class for first order time dependent operators.
-/** Operator of the form: (x,t) -> f(x,t), where k = f(x,t) generally solves the
-    algebraic equation F(x,k,t) = G(x,t). The functions F and G represent the
-    _implicit_ and _explicit_ parts of the operator, respectively. For explicit
-    operators, F(x,k,t) = k, so f(x,t) = G(x,t). */
+/** Operator of the form: (u,t) -> k(u,t), where k generally solves the
+    algebraic equation F(u,k,t) = G(u,t). The functions F and G represent the
+    _implicit_ and _explicit_ parts of the operator, respectively.
+
+    A common use for this class is representing a differential algebraic
+    equation of the form \f$ F(y,\frac{dy}{dt},t) = G(y,t) \f$. Consider an
+    ordinary differential equation of the form \f$ M \frac{dy}{dt} = g(y,t) \f$,
+    where F and G are defined as one of the following:
+    1. F(u,k,t) = k and G(u,t) = inv(M) g(u,t)
+    2. F(u,k,t) = M k and G(u,t) = g(u,t)
+    */
 class TimeDependentOperator : public Operator
 {
 public:
    enum Type
    {
-      EXPLICIT,   ///< This type assumes F(x,k,t) = k, i.e. k = f(x,t) = G(x,t).
+      EXPLICIT,   ///< This type assumes F(u,k,t) is linear in k.
```

Comment:
I don't like the change of meaning here -- this is not backward compatible.

If we want to have a type where `F` is linear in `k` (i.e. in `du/dt`) then it should be a new separate type.

Reply:
I have to admit that I didn't quite understand the use of `EXPLICIT` vs `IMPLICIT`... is there an example somewhere that differentiates between the two?

In solving $M \frac{dy}{dt} = g(y,t)$, is the related `TimeDependentOperator` considered `EXPLICIT`?

Reply:
For what it's worth, I had assumed the meaning of `EXPLICIT` vs `IMPLICIT` was equivalent to the PETSc `TSEquationType` documented at https://petsc.org/release/manualpages/TS/TSEquationType/

![image](https://github.com/mfem/mfem/assets/10469466/82ad8ada-cab8-47e6-b348-1740f8cc1895)

Reply:
The idea for the purpose of the `Type` is not well explained: the idea is that classes that use `TimeDependentOperator`s (like time-steppers) can query a given `TimeDependentOperator` about the form of ODE it represents. For example, if a user defines a derived class in which the form `du/dt = f(u,t)` is used, then they should setup their class to return `EXPLICIT` which will also imply that there is no difference between `Mult()` and `ExplicitMult()`, in particular. Of course, it is not wrong for the user class to return `IMPLICIT` (which is more general than `EXPLICIT`) in this case, however, the class that uses such `TimeDependentOperator` may try (if it makes decisions based on the returned type) to treat the ODE in its general (`IMPLICIT`) form and call methods designed for that case like `ImplicitMult()`, `ExplicitMult()`, or `GetImplicitGradient()`.

This is the intended use for the "type". However, I'm not really sure if any of the time steppers actually make use of it. Instead they just call the methods they want, assuming they are implemented and potentially assuming one of the "types" without actually checking. In some cases, not checking is perfectly fine, e.g. if the time stepper uses methods designed to work in all cases like `Mult()` or `ImplicitSolve()`.

Reply:
> I have to admit that I didn't quite understand the use of `EXPLICIT` vs `IMPLICIT`... is there an example somewhere that differentiates between the two?
> 
> In solving Mdydt=g(y,t), is the related `TimeDependentOperator` considered `EXPLICIT`?

It depends on the how the user-defined subclass of `TimeDependentOperator` treats the ODE: if `ExplicitMult()` returns `g(y,t)` then the class should return `IMPLICIT`; if `ExplicitMult()` returns `inv(M) g(y,t)` then it can (and typically should) return `EXPLICIT` but it is not wrong to return `IMPLICIT`.

Note that independent of the "type", `Mult()` should always return `inv(M) g(y,t)`, if it is implemented (some time-steppers may just not use `Mult()` so it will be fine to leave it undefined.)

Reply:
Thank you for clarifying, I think I have a good handle on things now.  

>Note that independent of the "type", Mult() should always return inv(M) g(y,t),

Ah ok, this was where I think I was getting confused: currently `ARKStepSolver` will assume `Mult` returns $g(y,t)$ if there has been a call to `ARKStepSolver::UseMFEMMassSolver`.  With the change I proposed to have `ARKStepSolver` instead use `ExplicitMult` to get $g(y,t)$ when there has been a call to `ARKStepSolver::UseMFEMMassSolver`, this will all be consistent.

I'll push a commit with an iteration of changes.

Reply:
I just pushed a commit that effectively reverted this line (but using the new notation).

Reply:
I believe this is resolved, please feel free to re-open this thread if it is not.